### PR TITLE
[REVIEW] Update `copy-on-write` with `branch-23.02` changes

### PR DESCRIPTION
**Pull request by galipremsagar**
## Description
This PR updates the `copy-on-write` branch with `get_ptr` & `data_array_view` APIs.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: python/cudf/cudf/tests/test_dataframe.py [hunk 1]
```diff
@@ -2660,12 +2660,20 @@ def query_GPU_memory(note=""):
...
 
     assert (
-        cudaDF.index._values.data_array_view.device_ctypes_pointer
-        == cudaDF["col0"].index._values.data_array_view.device_ctypes_pointer
+        cudaDF.index._values.data_array_view(
+            mode="write"
+        ).device_ctypes_pointer
+        == cudaDF["col0"]
+        .index._values.data_array_view(mode="write")
+        .device_ctypes_pointer
```
Comment (wence-) [inline python/cudf/cudf/tests/test_dataframe.py:2668]: I know it's only tests, but do we need to provoke copies here?

File: python/cudf/cudf/tests/test_spilling.py [hunk 2]
```diff
@@ -519,6 +519,7 @@ def test_get_rmm_memory_resource_stack():
 def test_df_transpose(manager: SpillManager):
     df1 = cudf.DataFrame({"a": [1, 2]})
     df2 = df1.transpose()
+    # import pdb;pdb.set_trace()
```
Comment (wence-) [inline python/cudf/cudf/tests/test_spilling.py:522]: Remove commented code.

File: python/cudf/cudf/utils/applyutils.py [hunk 3]
```diff
@@ -143,17 +143,20 @@ def run(self, df, **launch_params):
...
-                v: df[k]._column.data_array_view
+                v: df[k]._column.data_array_view(mode="write")
                 for (k, v) in self.incols.items()
             }
         else:
-            inputs = {k: df[k]._column.data_array_view for k in self.incols}
+            inputs = {
+                k: df[k]._column.data_array_view(mode="write")
+                for k in self.incols
+            }
```
Comment (wence-) [inline python/cudf/cudf/utils/applyutils.py:153]: Given these are inputs to a kernel launch that don't escape this function, can they not be `read` mode accesses?

File: python/cudf/cudf/utils/applyutils.py [hunk 4]
```diff
@@ -174,7 +177,7 @@ def run(self, df, **launch_params):
             )
             if out_mask is not None:
                 outdf._data[k] = outdf[k]._column.set_mask(
-                    out_mask.data_array_view
+                    out_mask.data_array_view(mode="write")
```
Comment (wence-) [inline python/cudf/cudf/utils/applyutils.py:180]: So `out_mask` is just created, so this is just an indication that we are taking ownership of this object, I guess.

File: python/cudf/cudf/utils/applyutils.py [hunk 5]
```diff
@@ -213,11 +216,11 @@ def launch_kernel(self, df, args, chunks, blkct=None, tpb=None):
     def normalize_chunks(self, size, chunks):
         if isinstance(chunks, int):
             # *chunks* is the chunksize
-            return column.arange(0, size, chunks).data_array_view
+            return column.arange(0, size, chunks).data_array_view(mode="write")
         else:
             # *chunks* is an array of chunk leading offset
             chunks = column.as_column(chunks)
-            return chunks.data_array_view
+            return chunks.data_array_view(mode="write")
```
Comment (wence-) [inline python/cudf/cudf/utils/applyutils.py:223]: Could we avoid going through a `Column` and just directly make cupy arrays in these two cases?

**Review by wence- — COMMENTED** [review]
Thanks, looks like a nice improvement. As a consequence, I have a few queries about places where I think we don't need to access `Column` arrays in `write` mode (and some where I think we can bypass creating a column at all, since we do so just to immediately turn into a cupy array).

File: python/cudf/cudf/_lib/column.pyx [hunk 6]
```diff
@@ -538,52 +528,61 @@ cdef class Column:
...
+                    # This is an optimization of the most common case where
+                    # from_column_view creates a "view" that is identical to
+                    # the owner.
+                    column_owner and
+                    isinstance(data_owner, SpillableBuffer) and
+                    # We check that `data_owner` is spill locked (not
+                    # spillable) and that its pointer is the same
+                    # as `data_ptr` _without_ exposing the buffer
+                    # permanently (calling get_ptr with a
+                    # dummy SpillLock).
```
[the remark this answers is not in the document]
Comment (galipremsagar) [reply python/cudf/cudf/_lib/column.pyx:557]: These will be updated once https://github.com/rapidsai/cudf/pull/12564 is merged

File: python/cudf/cudf/tests/test_dataframe.py (hunk 1 above)
Comment (galipremsagar) [reply python/cudf/cudf/tests/test_dataframe.py:2666]: Nope, not needed. Updated the test.

File: python/cudf/cudf/utils/applyutils.py (hunk 3 above)
Comment (galipremsagar) [reply python/cudf/cudf/utils/applyutils.py:153]: You are correct, not needed here too.

Comment (galipremsagar) [reply python/cudf/cudf/utils/applyutils.py:153]: Fixed it.

File: python/cudf/cudf/utils/applyutils.py (hunk 4 above)
Comment (galipremsagar) [reply python/cudf/cudf/utils/applyutils.py:180]: Right

File: python/cudf/cudf/utils/applyutils.py (hunk 5 above)
Comment (galipremsagar) [reply python/cudf/cudf/utils/applyutils.py:223]: Yup, made the change to use cupy arrays directly.

File: python/cudf/cudf/_lib/column.pyx [hunk 7]
```diff
@@ -223,7 +214,7 @@ cdef class Column:
                 value = value._get_readonly_proxy_obj
             if value.__cuda_array_interface__["typestr"] not in ("|i1", "|u1"):
                 if isinstance(value, Column):
-                    value = value.data_array_view
+                    value = value.data_array_view(mode="write")
                 value = cp.asarray(value).view('|u1')
```
[the remark this answers is not in the document]
Comment (galipremsagar) [reply python/cudf/cudf/_lib/column.pyx:218]: Looking into it..

File: python/strings_udf/strings_udf/tests/test_string_udfs.py [hunk 8]
```diff
@@ -1,4 +1,4 @@
-# Copyright (c) 2022, NVIDIA CORPORATION.
+# Copyright (c) 2022-2023, NVIDIA CORPORATION.
```
Comment (galipremsagar) [inline python/strings_udf/strings_udf/tests/test_string_udfs.py:1]:
```suggestion
# Copyright (c) 2022, NVIDIA CORPORATION.
```

File: python/strings_udf/strings_udf/tests/test_string_udfs.py [hunk 9]
```diff
@@ -10,6 +10,7 @@
 
 import cudf
 import rmm
+from cudf.core.buffer import acquire_spill_lock
```
Comment (galipremsagar) [inline python/strings_udf/strings_udf/tests/test_string_udfs.py:13]:
```suggestion
```

**Pull request merged**